src/textual/widgets/_option_list.py

```diff
@@ -0,0 +1,943 @@
+        """
+        self._needs_refresh_content_tracking = True
+        self._needs_to_scroll_to_highlight = rescroll_to_highlight
+        self.check_idle()
+
+    def on_idle(self) -> None:
+        """Perform content tracking data refresh when idle."""
+        self._refresh_content_tracking()
+        if self._needs_to_scroll_to_highlight:
+            self._needs_to_scroll_to_highlight = False
+            self.scroll_to_highlight()
+
+    def watch_show_vertical_scrollbar(self) -> None:
+        """Handle the vertical scrollbar visibility status changing.
+
+        `show_vertical_scrollbar` is watched because it has an impact on the
+        available width in which to render the renderables that make up the
+        options in the list. If a vertical scrollbar appears or disappears
+        we need to recalculate all the lines that make up the list.
+        """
+        self._request_content_tracking_refresh()
+
+    def on_resize(self) -> None:
+        """Refresh the layout of the renderables in the list when resized."""
+        self._request_content_tracking_refresh(rescroll_to_highlight=True)
+
+    def on_mouse_move(self, event: MouseMove) -> None:
+        """React to the mouse moving.
+
+        Args:
+            event: The mouse movement event.
+        """
+        self._mouse_hovering_over = event.style.meta.get("option")
+
+    def on_leave(self) -> None:
+        """React to the mouse leaving the widget."""
+        self._mouse_hovering_over = None
+
+    def on_click(self, event: Click) -> None:
+        """React to the mouse being clicked on an item.
+
+        Args:
+            event: The click event.
+        """
+        clicked_option = event.style.meta.get("option")
+        if clicked_option is not None:
+            self.highlighted = clicked_option
+            self.action_select()
+
+    def _make_content(self, content: NewOptionListContent) -> OptionListContent:
+        """Convert a single item of content for the list into a content type.
+
+        Args:
+            content: The content to turn into a full option list type.
+
+        Returns:
+            The content, usable in the option list.
+        """
+        if isinstance(content, (Option, Separator)):
+            return content
+        if content is None:
+            return Separator()
+        return Option(content)
+
+    def _clear_content_tracking(self) -> None:
+        """Clear down the content tracking information."""
+        self._lines.clear()
+        self._spans.clear()
+        # TODO: Having the option ID tracking be tied up with the main
+        # content tracking isn't necessary. Can possibly improve this a wee
+        # bit.
+        self._option_ids.clear()
+
+    def _refresh_content_tracking(self, force: bool = False) -> None:
+        """Refresh the various forms of option list content tracking.
+
+        Args:
+            force: Optionally force the refresh.
+
```

Comment:
Raises: DuplicateID missing.

Reply:
And more importantly on `add_option` too, where the user will actually see it.